### ISSUE TITLE
Incorrect deserialization for `BigDecimal` numbers

### ISSUE THREAD

**Issue by cristian-mocanu-mob** [FasterXML/jackson-databind#1770]
If the number in JSON cannot be represented as Double and ``DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS`` is enabled, big numbers are returned as Infinity.

The test below should pass:
```java
        final ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);

        final JsonNode jsonNode = objectMapper.readTree("7976931348623157e309");
        Assert.assertTrue(jsonNode instanceof NumericNode);

        // the following fails with NumberFormatException, because jsonNode is a DoubleNode with a value of POSITIVE_INFINITY
        Assert.assertTrue(jsonNode.decimalValue().compareTo(new BigDecimal("7976931348623157e309")) == 0);
```

The problem is reproducible in the latest released version, 2.9.1.

**Comment by cowtowncoder** [comment]
Thank you for reporting this, does sound like a bug.

**Comment by cowtowncoder** [comment]
Hmmh. This may be tricky to handle... problem being that detection of `NaN` case is triggered, and since value itself does round up to positive infinity (when attempted with `double`), value is not coerced as `BigDecimal` as it has no NaN options.

Will add failing test, see if I can figure out a way to handle this as expected.

**Comment by GedMarc** [comment]
instead of .decimal, please use .double

By default BigDecimal comes with a scale of 0, so decimal will only ever print out "0"

```

		ObjectMapper mapper = new ObjectMapper()
				                      .enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);

		System.out.println(BigDecimal.ZERO.compareTo(BigDecimal.valueOf(0.0)) == 0); //true
		System.out.println(BigDecimal.ZERO.compareTo(BigDecimal.valueOf(0)) == 0); //true

		BigDecimal a = new BigDecimal("0.00");
		BigDecimal b = new BigDecimal("0.0");
		BigDecimal c = new BigDecimal("0");

		if(a.doubleValue()==BigDecimal.ZERO.doubleValue()) {
			System.out.println("a equals");
		}

		if(b.doubleValue()==BigDecimal.ZERO.doubleValue()) {
			System.out.println("b equals");
		}

		if(c.doubleValue()==BigDecimal.ZERO.doubleValue()) {
			System.out.println("c equals");
		}

		JsonNode treeMode = mapper.readTree("{\"value\": 0.0}");
		Object value = treeMode.get("value").doubleValue();
		Object value2 = treeMode.get("value").decimalValue();
		System.out.println("value: " + value.getClass().getName() + " " + value.toString());
		System.out.println("value decimal : " + value2.getClass().getName() + " " + value2.toString());

		TypeReference mapOfStringObject = new TypeReference<Map<String, Object>>() {};
		Map<String, Object> map = mapper.readValue("{\"value\": 0.0}", mapOfStringObject);
		value = map.get("value");
		System.out.println("value: " + value.getClass().getName() + " " + value.toString());

```

**Comment by cowtowncoder** [comment]
@GedMarc maybe this comment is for different issue... ?

**Comment by GedMarc** [comment]
Oh Geez! This is for the double deserialization as float into big decimal!
This week has taken it out of me sigh, apologies!

**Comment by cowtowncoder** [comment]
For context, looks like fix for #1028 possibly caused this. So code that should coerce value results in parser trying to determine if it got a `NaN` and... in a way it is, wrt range of `Double`. But it would not be for `BigDecimal`.

**Comment by cowtowncoder** [comment]
@pjfanning This issue is something where I was hoping lazy parsing might work. I do not remember all the details on exact failure but seems like this could maybe be revisited to see it was now solvable in 2.16.
